core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1055,7 +1055,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   /** Build the union of a list of RDDs. */
   def union[T: ClassTag](rdds: Seq[RDD[T]]): RDD[T] = {
     val partitioners = rdds.flatMap(_.partitioner).toSet
-    if (partitioners.size == 1) {
+    if (rdds.forall(_.partitioner.isDefined) && partitioners.size == 1) {
```

Comment:
Yeah I like this. I suppose that the pre-existing condition already caught the empty RDD case, which `PartitionerAwareUnionRDD` will reject. Although symmetry between this check and the following one would be nice I don't think it's important. This looks correct since clearly `PartitionerAwareUnionRDD` intends to operate only on RDDs with partitioners.